Add counter

totokartonio opened the pull request:
1. Добавлена чистая функция, считающая количество значений в массиве, соответствующих переданному ей фильтру.
2. Добавлен счетчик активных событий, основывающийся на вышеуказанной функции.

Commit:
Clear completed

1. Добавлена чистая функция для очистки массива по передаваемому фильтру
2. Создана кнопка и настроена логика удаления выполненных задач.

Commit:
Add counter

1. Добавлена чистая функция, ссчитающая количество значений в массиве, соответствующих переданному ей фильтру.
2. Добавлен счетчик активных событий, основывающийся на вышеуказанной функции.

Commit:
Revert "Add counter"

This reverts commit 49f159683a437e17051a7b5a60a3dc7ffda69c60.

Commit:
Add counter

1. Добавлена чистая функция, ссчитающая количество значений в массиве, соответствующих переданному ей фильтру.
2. Добавлен счетчик активных событий, основывающийся на вышеуказанной функции.

totokartonio committed:
Rebase counter-active

---------

Co-authored-by: Anton Efimchenko <efimchenkoa@MacBook-Pro-Anton.local>

totokartonio committed:
Merge branch 'main' into counter-active

Commit:
Take care of conflicts

Commit:
Render utilities

1. Теперь utilities появляются вместе с первой задачей! Реализовано через CSS и псевдо-класс :empty

iffjey commented on styles/style.css:
```diff
@@ -53,6 +53,12 @@ body {
+  justify-content: center;
+}
+
+.task-list:empty + .utilities {
```
good!

iffjey commented on script.js:
```diff
@@ -101,6 +120,8 @@ const renderList = () => {
 
   listOfTasks.replaceChildren(fragment);
+
+  counterUpdate();
```
ну вот повторюсь, renderList уже делает все что нужно, этого здесь не должно быть

почему бы просто не сделать функцию `render`, которая пусть не оптимально, но будет держать в себе все нужные перерисовки

```
const render = () => {
  renderList();
  renderCounter();
}
```

iffjey replied:
ну или на крайний случай просто рядышком бы их поставил во все места, где нужно

можешь ещё например придумать какой-то параметр, который будет определять, что нужно перерисовать

```
const render = (areas = ['list', 'counter']) => {
  areas.forEach((area) => {
     const renderArea = SOME_MAP_OF_RENDER_FUNCTIONS[area];
     renderArea();
  })
} 
```

iffjey commented on script.js:
```diff
@@ -24,6 +25,22 @@ const createIncrementingIdGetter = () => {
 const getTaskId = createIncrementingIdGetter();
 
+//Функция для подсчета значений в массиве согласно фильтру
+const countElementsInArray = (arr, filterKey) => {
+  let i = 0;
+  arr
+    .filter(el => !el[filterKey])
+    .forEach(() => {
+      i++;
+    });
+  return i;
+};
```
ну опять же лучше, конечно сделать эту функцию чуть более конкретной `countCompleted` / `getCompletedCount`

имплементация чуть переусложнена — вот о чем можно подумать
1. Есть ли какой-то метод массива который держит в себе промежуточный результат, и может выдать цифру прям на выходе
2. Если ты уже отфильтровал массив и в нём остались только completed таски нужно ли нам по ним идти forEach или некоторое свойство массива нам может помочь?
3. Когда ты ответишь на два предыдущих вопроса, стоит подумать какой из способов все-таки будет оптимальнее?

iffjey commented on index.html:
```diff
@@ -16,15 +16,16 @@
           <option value="active">Active</option>
         </select>
         <button type="button" id="clear-completed">Clear all completed</button>
+        <p><span id="counter">0</span> items left!</p>
```
тут в теории ещё можно было бы задуматься про плюрализацию. Например, у тебя будет `1 items left`

Commit:
Create render function

1. Добавил плюрализацию в counter, теперь остальной текст также в span и меняется в зависимости от значения counter.
2. Переработал ccountCompleted: я нашел два метода массива, которые могут справляться с этой работй: .length и .reduce. На мой взгляд
оптимальнее .length, так как согласно документации if code becomes clearer when using other array methods, developers must weigh the readability tradeoff against the other benefits of using reduce().
Оставил .reduce для проверки исполнения.
3. Объединил функции render в мапу.

Commit:
Correct issues

Commit:
Fix filter bug

Для исправления бага, при котором если выбрано поле фильтра, а подходящих задач нет, то
DOM становился пустым, а utilities исчезали, теперь поля фильтра блокируютяс, пока не
появятся подходящие задачи. Для этого переработан counter, теперь это объект с
двумя функциями, counter.active() & counter.completed(). Сама блокировка
производится функцией renderFilter, которая в свою очередь стала частью
общей функции render. Был замечен другой баг: если у нас выбрано одно из условий фильтра,
и выбранные задачи исчезают (удаляются, меняют статус), в итоге DOM становится пустым и
utilities снова исчезают. Чтобы этого избежать, введена функция checkFilter(), которая
когда количество задач становится равно 0 и условие фильтра не All -- сбрасывает фильтр
до значения All функцией resetFilter(), что вместе с функцией render() позволяет снова
высветить на экране задачи, ранее не соответствующие условию фильтра.
Для улучшения логики процесса resetFilter() теперь срабатывает и при введении новой задачи:
какой бы фильтр не был выставлен до этого, он сбросится до All, чтобы не возникало ситуаций,
когда у насс отображаются только задачи Completed, а новая задача исчезает в недрах массива,
не появляясь на экране.

Commit:
End the PR

1. Нестрогие неравенста изменены на строгие.
2. В renderCounter упрощено создание строки.
3. Изменения checkBoxHandler теперь тригерят только render(['counter']).
4. Для utilities убрана связка .task-list:empty + .utilities на новый класс
.hidden. Манипуляция с новым классом производит функция renderClassList().

Удалены:
1. id у опций фильтра.
2. renderFilter(), checkFilter(): их задача теперь ыполняется renderClassList().
3. resetFilter(): будет переработана и добавлена позже.

Commit:
Correct mistakes
